### ISSUE TITLE
Sane UTF-8 in `cord`: `+teff`, `+sane`, and `+ruth`

### ISSUE THREAD

**Issue by ashelkovnykov** [urbit/urbit#6473]
During the JSON jet PRs (#5566, #6463), I introduced code to fix invalid UTF-8 parsing. This [started a discussion](https://github.com/urbit/urbit/pull/5566#discussion_r801248377) about whether or not it was correct to parse a `cord` as a UTF-8 string, since a `cord` is by definition supposed to be valid UTF-8. Some solution was required, since at the time the JSON parser would accept invalid UTF-8. However, the discussion is still ongoing.

One solution is to parse any JSON string to a `cord` and then sanity check that it's valid UTF-8 using `+sane`:
```
;~(pose (sear |=(a=@t ?.(((sane %t) ~ `a)) (cook crip ;~(less doq bas prn)) (cook crip esca))
```
There's a branch in `+sane` that is meant to check for `@t` sanity, but it's not currently implemented. This would be the correct place to insert jetted UTF-8 verification logic.

As a side note: wider sanity checking using `+ruth` (which uses `+sane` under the hood) is meant to be enabled, but blocked by two issues:
1. As mentioned above, `+sane` is incomplete
2. Arvo contains a large number of aura violations, as auras are not strictly validated (in fact, it's possible to create your own auras, so long as they nest properly and are used appropriately)

However, [another issue](https://github.com/urbit/urbit/pull/6463#pullrequestreview-1381864630) is that `cord` is actually a stricter type than just valid UTF-8: `cord` rejects newline characters. This means that the `+teff` used by `cord` is invalid for sanity checking JSON UTF-8 (it's too strict).

The current state is that the JSON parser in `+de:json:html` defines its own `+teff` and two arms for verifying UTF-8 sanity: `+sune` and `+sung`. In a perfect world,:
- There would be no difference between the `+teff` used by `hoon.hoon` and the one used by the JSON parser
- `+sane` would use `+sune` and `+sung`, which would be jetted, to sanity check UTF-8
- `+ruth` would call `+sane`
- Code that needs sane strings would call `+ruth`

**Comment by tinnus-napbus** [comment]
in what sense does `cord` reject newlines? Both the string literal parser & +sane accept them & always have done afaik

```
> 'foo\0abar\0a'
'foo\0abar\0a'
> ((sane %t) 'foo\0abar\0a')
%.y
```

**Comment by joemfb** [comment]
`(sane %t)` and `+teff` both reject `0-9` and `11-31`; they allow "line feed" (`\\n`), but reject (other?) control characters including "carriage return" and "tab". (I think there was an early attempt to exclude line feed as well, or maybe just a stated intention to do so. I'll post what I find.)

**Comment by tinnus-napbus** [comment]
oh right yeah:

```
  ?:  &((lth cur 32) !=(10 cur))  |
```

I feel like it should allow any characters

**Comment by mikolajpp** [comment]
> `(sane %t)` and `+teff` both reject `0-9` and `11-31`; they allow "line feed" (`\\n`), but reject (other?) control characters including "carriage return" and "tab". (I think there was an early attempt to exclude line feed as well, or maybe just a stated intention to do so. I'll post what I find.)

I was interested in finding about this and after some digging stumbled upon this 10 years old comment:  https://github.com/urbit/urbit/issues/229#issuecomment-68154003.

Meanwhile, my UIP-0107 on Aura Renovation should soon be finalized, see current @c and @t spec here: https://github.com/mikolajpp/UIPs/blob/aura-renovation/UIPS/UIP-0107.md#specification-supplement.  

I have considered the idea of exiling the newline, but I don't think it makes sense: unlike other ASCII control codes, the newline has consistent meaning across all software, which either prints it escaped, or performs a line break.
While there could be a hypothetical advantage of structural sharing of lines of text across the system, this is already provided  by storing text as a `wain`. Removing newline from cords would also break the current parsing framework.
